src/sst/elements/vanadis/decoder/vriscv64decoder.h

```diff
@@ -1107,16 +1105,29 @@ class VanadisRISCV64Decoder : public VanadisDecoder
                       } break;
                       default:
                       {
+                        using namespace Zicntr;
                         uint64_t csrNum = uimm64 & 0xfff;
                         switch ( csrNum ) {
-                            case 0xc00:
-                            {
-                                if ( 0 == rs1 ) {
-                                    auto thread_call = std::bind(&VanadisRISCV64Decoder::getCycleCount, this);
-                                    bundle->addInstruction( new VanadisSetRegisterByCallInstruction<int64_t>( ins_address, hw_thr, options, rd, thread_call));
-                                    decode_fault = false;
-                                }
-                            } break;
+                            case 0xc00: // RDCYCLE
+                                bundle->addInstruction( new VanadisReadCounterInstruction( CYCLE, ins_address, hw_thr, options, rd ) );
+                                decode_fault = 0 != rs1;
+                                break;
+                            
+                            case 0xc01: // RDTIME
+                                bundle->addInstruction( new VanadisReadCounterInstruction( TIME, ins_address, hw_thr, options, rd ) );
+                                decode_fault = 0 != rs1;
+                                break;
+                            
+                            case 0xc02: // RDINSTRET
+                                bundle->addInstruction( new VanadisReadCounterInstruction( INSTRET, ins_address, hw_thr, options, rd ) );
+                                decode_fault = 0 != rs1;
+                                break;
+
+                            case 0xc80: // RDCYCLEH                            
+                            case 0xc81: // RDTIMEH                            
+                            case 0xc82: // RDINSTRETH
+                                output->verbose( CALL_INFO, 16, 0, "riscv64 does not support Zicntr [H] suffix" );
```

Comment:
Should this kill the simulation?

Reply:
I assumed that it would be appropriate for this to be considered a `decode_fault` with the verbose output giving a little more information up-front. I'm not particularly tied to any handling here. I'm not sure what the specification has to say about encountering instructions that aren't implemented/able given the implementation's ABI.

Reply:
I think this is a reasonable choice; I am also not sure what the specification has to say. The Vanadis config controls whether a `decode_fault` results in a kill.